packages/astro/src/actions/runtime/virtual/server.ts

```diff
@@ -60,7 +60,8 @@ export function defineAction<
 			: getJsonServerHandler(handler, inputSchema);
 
 	async function safeServerHandler(this: ActionAPIContext, unparsedInput: unknown) {
-		if (typeof this === 'function') {
+		// The ActionAPIContext should always contain the `params` property
+		if (typeof this === 'function' || !('params' in this)) {
```

Comment:
Good with this as a fix, but would be nice to have a `Symbol` we can check for to verify that an `ActionAPIContext` was passed. Could be abstracted to an `isActionAPIContext()` type guard

Reply:
Good suggestion! Thank you